### PR TITLE
[IMP] account_tax_cash_basis: add tests in order to validate the correct

### PR DESCRIPTION
**Pull request by JulioSerna**
amount about the creation of the tax cash basis  when there is rate difference
between date invoice and date payment

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account_tax_cash_basis/__init__.py [hunk 1]
```diff
@@ -2,3 +2,4 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import models
+from . import tests
```
Comment (xmo-odoo) [inline addons/account_tax_cash_basis/__init__.py]: Don't import test modules from the main modules.

File: addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+       Invoice | 1.2834 | Not yet created
+       Payment | 1.5289 | 10.47
+        """
+        invoice_id = self.create_invoice_usd(time.strftime('%Y') + '-01-01')
+        self.create_payment(invoice_id, time.strftime('%Y') + '-07-15', 116)
+        move_ter = self.account_move_line_model.search(
+            [('account_id', '=', self.account_tax_cash_basis.id)])
+        ter_balance = (
+            sum(move_ter.mapped('credit')) - sum(move_ter.mapped('debit')))
+        self.assertEquals(round(ter_balance, 2), 10.47)
```
Comment (xmo-odoo) [inline addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py]: Avoid builtin round, use `float_round` instead.

File: addons/account_tax_cash_basis/__init__.py (hunk 1 above)
Comment (JulioSerna) [inline addons/account_tax_cash_basis/__init__.py]: change applied

File: addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py (hunk 2 above)
Comment (JulioSerna) [inline addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py]: Change applied

**Comment by tivisse** [comment]
Hello @JulioSerna 

I see that the test you wrote is red on runbot. 

What is the purpose of the test ? 
- Demonstrate that the current behavior is wrong ? In that case could you explain what should be the correct behavior with a valid and complete business case.
- Add a test to validate the current behavior ? In that case your test is red an you should update it accordingly. 

Waiting your input,

Yannick.

**Comment by JulioSerna** [comment]
@tivisse 

The purpose is the first, Demonstrate that the current behavior is wrong
and this is the validate case,

e.g
using date demo to rate and amount

date  |  rate      | inverse rate
------|---------|-----------
12/31/2009 18:00:00|1.2834|0.7791803023
06/05/2017 19:00:00|1.5289|0.6540650141

1. set currency company EUR
2. Create tax 16% with `use_cash_basis` and set account `cash_basis_account` 
3. The account set in `cash_basis_account` it  will be **Tax effectively receivable**
4. Create invoice with date  **12/31/2009 18:00:00** and currency **USD**
5. Create invoice line price unit 100 and tax 16%
6. validate invoice
7. create payment by **$116** with date **06/05/2017 19:00:00** and currency **USD** 
8. Reconcile invoice(apply payment should be by bank statement or payment directly from the invioce)

**Current Behavior**

The balance of the account **Tax effectively receivable** is **$12.47** I think this amount is because is getting the rate of the date invioce(**12/31/2009 18:00:00**) `(rate invioce)0.7791803023*16(tax usd) = $12.47`

**Expected Behavior**
The amount of the account **Tax effectively receivable** should be **$10.47** this is because the amount must be convert into date of the payment(**06/05/2017 19:00:00**) `rate(0.6540650141) * 16(tax usd) = $10.47`

File: addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                    'account.data_account_type_revenue').id)], limit=1).id,
+            'invoice_line_tax_ids': [(4, tax.ids)]
+        })
+
+        # validate invoice
+        invoice.compute_taxes()
+        invoice.action_invoice_open()
+        return invoice
+
+    def test_cash_basis_tax_rate_invioce_greater_than_payment(self):
```
Comment (smetl) [inline addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py]: rename invioce to invoice, same in other methods name

File: addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+       Payment | 1.5289 | 10.47
+        """
+        invoice_id = self.create_invoice_usd(time.strftime('%Y') + '-01-01')
+        self.create_payment(invoice_id, time.strftime('%Y') + '-07-15', 116)
+        move_ter = self.account_move_line_model.search(
+            [('account_id', '=', self.account_tax_cash_basis.id)])
+        ter_balance = (
+            sum(move_ter.mapped('credit')) - sum(move_ter.mapped('debit')))
+        self.assertEquals(
+            float_round(ter_balance, precision_digits=self.precision), 10.47)
```
Comment (smetl) [inline addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py]: In this case, the following entries are created:

INVOICE:

account|debit|credit
-----|-----|-----
Receivable|90.39|
Fixed||12.47
Product||77.92
Tot.|90.39|90.39

BANK:

account|debit|credit
-----|-----|-----
Bank|75.87|
Receivable||77.87
Tot.|77.87|77.87

BANK EXCHANGE:

account|debit|credit
-----|-----|-----
Exchange|14.52|
Receivable||14.52
Tot.|14.52|14.52

CASH:

account|debit|credit
-----|-----|-----
Fixed|12.47|
Inter.||12.47
Product|77.92|
Product||77.92
Tot.|90.39|90.39

CASH EXCHANGE:

account|debit|credit
-----|-----|-----
Fixed|2.0|
Inter.||2.0
Product|12.52|
Product||12.52
Tot.|14.52|14.52

That leads to following balances:

FIXED:

debit|credit
-----|-----
2.0|0.0

PRODUCT:

debit|credit
-----|-----
0.0|77.92

EXCHANGE:

debit|credit
-----|-----
14.52|0.0

BANK:

debit|credit
-----|-----
75.87|0.0

INTER.:

debit|credit
-----|-----
0.0|14.47

I agree there is something wrong here (in INTER./FIXED). In my opinion, the error comes from the CASH EXCHANGE that should be:

 account|debit|credit
-----|-----|-----
Inter.|2.0|
Fixed||2.0
Product|12.52|
Product||12.52
Tot.|14.52|14.52

to change FIXED:

debit|credit
-----|-----
0.0|2.0

and having INTER. like:

debit|credit
-----|-----
2.0|12.47

... that seems correct according to this test.

My question is about FIXED, it is correct to have 2.0 remaining credit in this case?

Comment (JulioSerna) [inline addons/account_tax_cash_basis/tests/test_account_tax_cash_basis.py]: Yes that is right

**Comment by qdp-odoo** [comment]
As explained in the ticket you opened in our support channel, I find the motivations of this PR not good enough: to me, it seems logical that the cash basis entry is made at the same rate as the invoice since you expect that a full payment would clear completely the interim tax account.

Using the same rate is the only way to achieve it, without going into a reconciliation process... which I'd like to avoid if not necessary. 

So please would you provide the legal source proving the current behavior is wrong? Even though, I'd be reluctant to apply that to all countries using cash basis (maybe it is only in Mexico??)

**Comment by nhomar** [comment]
@qdp-odoo 

I think we have a huge conceptual problem here.

- For méxico it is by law and we received at least 10 internal issues for this (as first part).
- For countries like Canada, this adjustment can be done periodically (but it continue being as @JulioSerna  explain as well).
- There are issues here in github related to other country like this one: https://github.com/odoo/odoo/issues/9972
- The name **Cash Basis** express "Based on the cash"  I do not know which other word in english use to explain the same.
- The legal support is [here](http://losimpuestos.com.mx/que-tipo-de-cambio-usar-en-una-factura/) [Read the very last paragraph which has the legal reference of the law resumed].

> Para determinar las contribuciones y sus accesorios se considerará el tipo de cambio a que se haya adquirido la moneda extranjera de que se trate y no habiendo adquisición, se estará al tipo de cambio que el Banco de México publique en el Diario Oficial de la Federación el día anterior a aquél en que se causen las contribuciones. Los días en que el Banco de México no publique dicho tipo de cambio se aplicará el último tipo de cambio publicado con anterioridad al día en que se causen las contribuciones.

> Cuando las disposiciones fiscales permitan el acreditamiento de impuestos o de cantidades equivalentes a éstos, pagados en moneda extranjera, se considerará el tipo de cambio que corresponda conforme a lo señalado en el párrafo anterior, referido a la fecha en que se causó el impuesto que se traslada o en su defecto cuando se pague.

**the complete law where this information [is here](https://www.oas.org/juridico/spanish/mesicic3_mex_anexo6.pdf)**

- Believe US we do not want to change this more than you but **conceptually** you can not use the term "Based on Something" and then use partially on the code "Based on other thing" I mean my friend, seriously, the one that must show a law is yourself not us.
image.

Why do we think it is generic?

- This rationale is globally accepted, the current behavior is **the exception to the concept** no the rule on the concept itself, I think we can do it on an l10n_mx_?? module if you want and benefit only one country but i think that can be wrong from a design perspective.

Thanks in advance for your support men.

cc @JulioSerna cc @alan196

**Comment by wtaferner** [comment]
@nhomar 
I think there are good reasons behind your use case, but I have to add and backup @qdp-odoo that in Europe and probably not only in Europe normally if you bill in another currency than your own you have to give the VAT amount in your local currency on the INVOICE in order that the buyer and the seller have the same amount to pay or to claim against the tax authority. If one of the two is using this and the other one the other method there will be great issues to keep the VAT in line.

So, conceptually it is the right thing to do IMHO as it is. I think the terms used should be clarified more in detail and maybe your use case should be made possible under another option at some point.

Honestly there are plenty of use cases which are not possible yet with Odoo Accounting, but I think time will improve it and in the meantime we have to help ourselves.

Just my point of view and five cents on this issue...

**Comment by nhomar** [comment]
@wtaferner @qdp-odoo 

With all due respect, but nobody here is showing any law that support the conceptual development, if at least you mention a link to the EU lay that is opposite to what we mentioned then I move this test and we make the development to an l10n_mx_payment_exchange_rate with no problem and we fix just for our customers.

>> So please would you provide the legal source proving the current behavior is wrong? Even though, I'd be reluctant to apply that to all countries using cash basis (maybe it is only in Mexico??)

Dude I answered this question and sadly no response on your side (neither a legal support to the current feature or the against our argument).

>>Using the same rate is the only way to achieve it, without going into a reconciliation process... which I'd like to avoid if not necessary.

I think it is not.

We can create few extra move lines for the difference and simply reconcile them.

I mean I am not closed to do it at localization level, but I must to be so honest here, this is a bug that will kill in few months a lot of multicurrency implementations, increase the work done by the implementer and do not add any value to maintain as it is.

I hope you reconsidere this.

@JulioSerna 

In the meanwhile can you make a module called `l10n_mx_payment_exchange_rate` under vauxoo/mexico to solve this issue at Mexico level please.

This modules would do an extra account move with the difference post-processed in the exchange rate using a server action, this will be used internally to solve >v10.0 problems, and please propose the PR to master to have it as part of the localization for future usage?.

PS: @qdp-odoo in case the module is accepted, do you prefer a new option on taxes in master or an extra module as I proposed?

Regards.

**Comment by wtaferner** [comment]
@nhomar 
[Current EU helper about the taxable amount](https://ec.europa.eu/taxation_customs/business/vat/eu-vat-rules-topic/taxable-amount_en)

This is what I found and no, we are not in court and I guess we all have reasons, but for Europe KPMG was publishing under "Umrechnungskurs" of their [Tax News in 2012](https://www.kpmg.at/newsletter/files/TN_2012_23.pdf) that the EURO amount must be fixed on the invoice. So even if you are paid months later the EURO amount is fixed and is not newly calculated. I found [another source](http://www.tpa-group.at/de/publikationen/publikationen/newsletter/ust-die-rechnung-ab-01012013) of a tax service company which says that it must be fixed or at least told which conversion procedure is applied that both actors will have the same tax amount to pay and to claim. Anyway, IMHO the public should be always better off and the risk is to be taken by the companies, never by the public. 

This is not proving anything but showing that it is f*** complicated and obviously a lot of European companies work with Odoo and it was implemented like this on a conceptual level based on common sense in the EU...no one to blame, but most of the Enterprise users are Belgium and EU companies which is already changing and I hope that there will be better options in the future to make everyone happy ;-)

**Comment by nhomar** [comment]
Hello @wtaferner ---> THANK U! this kind of answers seariously help more than the opinionated ones.

I read carefully your very first [document](https://ec.europa.eu/taxation_customs/business/vat/eu-vat-rules-topic/taxable-amount_en):

I made a quick look for "Exchange" and the first answer was what you see on this image:

image

As you can see the exchange rate is:

  >> the exchange rate will be the latest recorded selling rate at the time the VAT becomes due on that EU country’s most representative exchange market(s).

Said this:

IF you own when you receive the cash is the date of the receipt of the cash if you own with the invoice then the rate is with the invoice.

What I can't find is the specific country that say "You own the vat when you receive the money but use the exchange rate when you invoiced".

LEt me explain why this is so inconsistent on countries that do not use Euros:

The Mexican exchange rate variance is here:

image

As you can see between the max exchange rate (invoiced date) and the min exchange rate (payment date) there is a difference of almost 15%.

The VAT is 16%, then it represent in total (just taking the worst case) 2.4% of the base amount.

Let's say you "export" agriculture merchandise (which the profit is +/-) 5% or technological reseller (which is more or less the same).

Because this "technical decision" your ERP is making loss almost 50% of you net income with no reasonable explanation (#realcaseBEENTHERE).

Then as consultant we can not allow this (IMHO) and with no fundament this is a really UGLY bug.

What is 100% true is that if there is **1 case** of the supported countries that my statement is false "please" give your opinion, because I think this must be **at least** configurable, **but not technically with the easiest way to do it even if it is a bug**.

@wtaferner 

About this:

> This is what I found and no, we are not in court and I guess we all have reasons, but for Europe KPMG was publishing under "Umrechnungskurs" of their Tax News in 2012 that the EURO amount must be fixed on the invoice. So even if you are paid months later the EURO amount is fixed and is not newly calculated.

Sorry my german is not good, what I need to ask is:

If You invoice 30 invoices and the sum of taxes is 100€, then you collect 15 Invoices summing the collected tax of 50€. How much do you own to the Fiscal entity that specific period, 100 or 50? --> This answer is important to know if we are talking about the same here.

**Comment by wtaferner** [comment]
@nhomar 
I do see your issues and as said I am pretty sure that you have reasons, but to make it short...your last question to be answered is...

If you do 30 invoices in a foreign currency (let's say USD) and you are situated in Austria the conversion is done at the invoice date into EUR and you will owe this amount to your government/tax authority.

So it is EUR 100,-- in total and will always stay the amount due, no matter if you use cash-tax-basis or anything else...but the day you owe them is different (so if you got paid half of it you will have to pay EUR 50 even if the exchange rate now says it is worth EUR 44 only or maybe EUR 80 instead)...as said, it must be in equilibrium...you can even do it within the same country to better understand the principle behind...an Austrian company (normal basis) is billing another Austrian company (cash-basis) in USD. So one company owes the tax directly in EUR on the day of the invoice and the other one can only claim the money in EUR on the day of the payment. If you do not freeze the amount the public or the tax authority has the risk of exchange which is circumvented by the principle that it must be converted on the day of the invoice, so it will be a zero sum game for the government and only the end customer is finally paying the VAT as he/she can not claim it from the tax authority (this is how VAT is setup and working IMHO).

So, as far as I can say the tax is not converted any time later than the invoice date, but yeah, it is difficult to talk about this while not sitting on the same table in RL and I am still not sure if we talk about the same as I have no clue about Mexican laws and practices :)

**Comment by wtaferner** [comment]
I think now I understand your concern, but this is IMHO the risk of foreign currency invoicing...you claim that the due date must be the exchange rate date, right?
But this is unreliable and prediction so you are unable to say what it will be -> Risk to be taken by the invoicing company...

**Comment by nhomar** [comment]
@wtaferner 

Sorry men to insist in order to not complicate, can you answer this specific question (forget for 1 second the exchange rate):

>> If You invoice 30 invoices and the sum of taxes is 100€, then you collect 15 Invoices summing the collected tax of 50€. How much do you own to the Fiscal entity that specific period, 100 or 50? --> This answer is important to know if we are talking about the same here.

**Comment by qdp-odoo** [comment]
fyi: I'm not far away, reading carefully your opinions and taking notes on possible actions to please everyone.

That being said, you've proven that it was an issue for your country, so I'm reopening the bug report

**Comment by gdgellatly** [comment]
@wtaferner actually your assertion on risk is incorrect.  It is when collecting tax on an invoice basis, that the invoicing company takes exchange risk on tax as they pay the tax at one rate and collect in another.  On a payment basis, a 25% tax is still 25%.

However on an invoice basis, a 25% tax could become effectively a 10% or 60% tax.

The funny thing is both you and @nhomar are correct even for the EU (and therefore @qdp-odoo not only is it shown as an issue for MX, but also EU if Wolfgang's text is correct).  And all your conflicting information makes sense.  

Most companies (maybe all) that engage in FX transactions in the EU pay tax on an invoice basis, therefore the tax is presumably technically due at date of invoice and accrued at that time.  Therefore it is the common practice and recommendation, despite it being strictly incorrect in some circumstances.  

However, if they were on a cash basis, then the tax amount becomes due on the date of payment, and as such the FX rate used is the payment date.  But I imagine again in practice this rule is ignored because most cash basis companies are small, their FX volumes small as well, exchange rates relatively stable and they are blissfully unaware of such rules and as the variances are tiny tax departments don't even bother to look let alone enforce.  But it doesn't make it right.

**Comment by gdgellatly** [comment]
BTW - I checked for NZ.  It took me a long time to find. There is no specific legislation for FX, except a court decision determining that the tax is converted to NZ currency at time of supply.

Time of supply is a little bit technical but usually refers to the earlier of invoice date or payment (even a part payment triggers the full liability).  Luckily our sales tax is a consumption tax rather than a VAT, making exports exempt and imports levied rather than taxed (so the rate is whatever customs tells you it is and paid in NZD), so in practice it would probably never come up.

**Comment by wtaferner** [comment]
@gdgellatly 
Thank you for trying to give a third perspective on this and I was not about to argue about being right or wrong at any time (my sources are KPMG and TPA, but even they just say it must be written on the invoice and does say nothing what is law or any stuff in detail...and even the text I provided from the official EU website is a helper not a  document of laws)...for me it was important to show that the point of view of tax authorities logically should not have a VAT GAP based on exchange rates. 
In the EU there are a lot of other issues with having a [VAT GAP](https://ec.europa.eu/taxation_customs/business/tax-cooperation-control/vat-gap_en) even between countries of the EU although it is supposed to be zero in a perfect world where no company is going bankrupt. Tax fraud is based on this, so my assumption is always that this should not be possible (I know a lot of money driven companies do not care about right or wrong on other levels than law)...
I do not even need to read any law b/c at some point it will become law when the law gap is closed and the rest is waste of time IMHO, but yeah, we try to build software to respect the laws and laws are very diverse and implemented in different levels of perfection, so it is clear for me that it is impossible to make it possible on a generic level.

What I do oppose to have in core are things which are built to satisfy a customer trying to use a law gap to his/her advantage and maybe even blame Odoo or the integrator for making it possible.

@nhomar 
In the end @qdp-odoo reopened the issue which is good and we are all richer by having multiple points of views and at least thought about it a bit...it was a pleasure to have looked a bit into it, but in the end it is no issue for me at the moment and just wanted to help to have this discussion opened up a bit.

**Comment by gdgellatly** [comment]
@wtaferner 
No tax gap theoretically exists due to FX translation and timing.  If the rule is "the date the tax becomes due" and the exchange rate is proscribed then the rule is the same for both ends of the transaction.  A tax gap may arise if both ends of the transaction have a different due date (e.g. one on payments and one on invoice), however, such a gap arises regardless of whether FX is involved, albeit in this case just a timing one.

Nonetheless, the rule you say from KPMG is in fact the law (more or less) in my country regardless of payments or invoice basis.  The same is true for the UK, where a VAT invoice is legally a sterling invoice, recorded in sterling and presented in Euro (or other currency) usually along with the sterling amounts (for exactly the fraud you described).  On reflection, I suspect that the rule is also true for the EU, and you merely found some badly worded documentation.  What is missing is that cash basis only applies for intracountry transactions, not exports or imports.

**Comment by wtaferner** [comment]
@gdgellatly 
Well, for export it is almost all the time 0% tax, so not an issue at all and import is based on customs and other workflows (landed costs) and is currently very individual even though in the EU.

As said, there is no rule, but they say that it must be written on the invoice either the foreign currency amount worth in EUR or at least a clear method how it is calculated that there is consequently no gap based on information.

You are dealing a lot with these issues I guess ;-) Most of the discussion I guess is based on misinterpretation of the written communication...so, I guess we all know what we are talking about in the end :-)

**Comment by gdgellatly** [comment]
@wtaferner These days not so much on sales tax, but once upon a time, sales tax for importer/exporters was a big part of my life.  But anyway, the export I was primarily referring to was inside the EU, e.g. CHF/EUR/GBP. (Probably about to become an even bigger issue)

**Comment by nhomar** [comment]
@qdp-odoo This is the vat exchange rate difference problem.

**Comment by tivisse** [comment]
Hello,

Thank you for your contribution but the branch is no longer supported. We only support the last 3 stable versions so no longer accepts patches into this branch.
We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

Regards

**Pull request closed without merging**